runtime/src/cost_model.rs

```diff
@@ -166,18 +171,26 @@ impl CostModel {
             // to keep the same behavior, look for builtin first
             if let Some(builtin_cost) = BUILT_IN_INSTRUCTION_COSTS.get(program_id) {
                 builtin_costs = builtin_costs.saturating_add(*builtin_cost);
-            } else {
-                let instruction_cost = self.find_instruction_cost(program_id);
-                trace!(
-                    "instruction {:?} has cost of {}",
-                    instruction,
-                    instruction_cost
-                );
-                bpf_costs = bpf_costs.saturating_add(instruction_cost);
+            } else if !compute_budget::check_id(program_id) {
+                bpf_costs = bpf_costs.saturating_add(DEFAULT_INSTRUCTION_COMPUTE_UNIT_LIMIT.into());
             }
             data_bytes_len_total =
                 data_bytes_len_total.saturating_add(instruction.data.len() as u64);
         }
+
+        // calculate bpf cost based on compute budget instructions
+        let mut budget = ComputeBudget::default();
+        let result = budget.process_instructions(
+            transaction.message().program_instructions_iter(),
+            true,
+            false, /*not support request_units_deprecated*/
+        );
+
+        // if tx contained user-space instructions and a more accurate estimate available correct it
+        if bpf_costs > 0 && result.is_ok() {
```

Comment:
not sure about "bpf_costs > 0" part. 

Maybe add a function like this to read requested cu, and use it straight at Ln 175?
```
    fn get_requested_compute_units(
        instructions: impl Iterator<Item = (&Pubkey, &CompiledInstruction)>,
    ) -> Option<u64> {
        let mut compute_budget = ComputeBudget::default();
        let _ = compute_budget
            .process_instructions(
                instructions,
                true, // don't reject txs that use request heap size ix
                true, // use default units per instruction
                true, // don't reject txs that use set compute unit price ix
            )
            .ok()?;
        Some(compute_budget.compute_unit_limit)
    }
```

Reply:
There's an issue with ComputeBudget::compute_unit_limit, it actually does calculate `num_non_compute_budget_instructions` and multiply that with `DEFAULT_INSTRUCTION_COMPUTE_UNIT_LIMIT`

This would estimate all votes at 200_000k for instance, so the idea to check for bpf_costs is to make sure that txs that only contain builtins do not suffer from this bad estimation.